pkg/proxy/backup.go

```diff
@@ -0,0 +1,444 @@
+package proxy
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/golang/protobuf/ptypes"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/net/context"
+
+	rpc "github.com/longhorn/longhorn-instance-manager/pkg/imrpc"
+
+	eclient "github.com/longhorn/longhorn-engine/pkg/controller/client"
+	rclient "github.com/longhorn/longhorn-engine/pkg/replica/client"
+	esync "github.com/longhorn/longhorn-engine/pkg/sync"
+	etypes "github.com/longhorn/longhorn-engine/pkg/types"
+	eutil "github.com/longhorn/longhorn-engine/pkg/util"
+	eptypes "github.com/longhorn/longhorn-engine/proto/ptypes"
+
+	"github.com/longhorn/backupstore"
+	bsutil "github.com/longhorn/backupstore/util"
+)
+
+func (p *Proxy) SnapshotBackup(ctx context.Context, req *rpc.EngineSnapshotBackupRequest) (resp *rpc.EngineSnapshotBackupProxyResponse, err error) {
+	log := logrus.WithFields(logrus.Fields{"serviceURL": req.ProxyEngineRequest.Address})
+	log.Debugf("Backing up snapshots %v to backup %v", req.SnapshotName, req.BackupName)
+
+	resp = &rpc.EngineSnapshotBackupProxyResponse{
+		ProxyEngineResponse: &rpc.ProxyEngineResponse{},
+		BackupId:            "",
+		Replica:             "",
+		IsIncremental:       false,
+	}
+	ackResponse(resp.ProxyEngineResponse, false)
+
+	newCtx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	for _, env := range req.Envs {
+		part := strings.SplitN(env, "=", 2)
+		if err := os.Setenv(part[0], part[1]); err != nil {
+			return resp, err
+		}
+	}
```

Comment:
This way will impact the environment variables of instance manager process? This will be different from the previous implementation: engineapi will start a process to run engine binary, so it can be launched with environment variables. (note: we only have one backup target now, so probably it will not have any impact, but still uncertain)

Probably you need to have a new GetBackupCredential which is getting info from the passed envs instead of os environment variables. The same for other places.

Reply:
https://github.com/longhorn/longhorn/issues/3977

Note, `SyncAgentServer` will still update the os env per request, for example, [BackupCreate](https://github.com/longhorn/longhorn-engine/blob/master/pkg/sync/rpc/server.go#L676-L701).